leader-election-impl/src/main/java/com/palantir/leader/proxy/AwaitingLeadershipProxy.java

```diff
@@ -159,18 +163,62 @@ protected Object handleInvocation(Object proxy, Method method, Object[] args) th
 
     private RuntimeException handleDelegateThrewException(
             LeadershipToken leadershipToken, InvocationTargetException exception) throws Exception {
-        if (exception.getCause() instanceof ServiceNotAvailableException
-                || exception.getCause() instanceof NotCurrentLeaderException) {
-            leadershipStateManager.handleNotLeading(leadershipToken, exception.getCause());
+        Throwable cause = exception.getCause();
+        if (cause instanceof SuspectedNotCurrentLeaderException) {
+            handleSuspectedNotCurrentLeader(leadershipToken, (SuspectedNotCurrentLeaderException) cause);
+        }
+        if (cause instanceof ServiceNotAvailableException || cause instanceof NotCurrentLeaderException) {
+            leadershipStateManager.handleNotLeading(leadershipToken, cause);
         }
         // Prevent blocked lock requests from receiving a non-retryable 500 on interrupts
         // in case of a leader election.
-        if (exception.getCause() instanceof InterruptedException
-                && !leadershipCoordinator.isStillCurrentToken(leadershipToken)) {
+        if (cause instanceof InterruptedException && !leadershipCoordinator.isStillCurrentToken(leadershipToken)) {
             throw leadershipCoordinator.notCurrentLeaderException(
-                    "received an interrupt due to leader election.", exception.getCause());
+                    "received an interrupt due to leader election.", cause);
+        }
+        Throwables.propagateIfPossible(cause, Exception.class);
+        throw new RuntimeException(cause);
+    }
+
+    private void handleSuspectedNotCurrentLeader(
```

Comment:
I see LSM's logic mostly as dealing with the delegate and token references and updating them correctly, while the logic of exception handling belongs here in ALP (e.g. the stuff around switching on the exception type). Though I agree that the naming of things with `handle` makes it easy to conflate the concepts.

I think I'll keep this pair as `handle...` since these are in relation to exceptions, while the rest are more reactions to not leading (or the delegate no longer being valid) so I'll find different verbs there - I've changed to `requestDelegateInvalidation` and `invalidateStateOnLostLeadership`.